File: examples/komegaExamples/kw_py/ts.py
import vsiputils as vsip

def VU_rowview(v, m, i):
   mattr=vsip.getattrib(m)
   vattr=vsip.getattrib(v)
   vattr.offset = mattr.offset + i * mattr.col_stride
   vattr.length = mattr.row_length
   vattr.stride = mattr.row_stride
   vsip.putattrib(v,vattr);

class Sim_ts(object):
    """ Simulate acoustic data with narrow band point sources from multiple
        directions and isotropic, band-limited noise.
    """
    def __init__(self,param):
        c=param['c']
        Dsens = param['Dsens']
        Nsens = param['Nsens']
        Nts = param['Nts']
        Fs = param['Fs']
        row=vsip.VSIP_ROW
        mem=vsip.VSIP_MEM_NONE
        state=vsip.VSIP_STATE_SAVE
        sym=vsip.VSIP_NONSYM
        rng=vsip.VSIP_PRNG
        L          = int(2 * Fs/(Nsens * Dsens/c) + Nts + 1)
        self.Nsim_freqs = param['Nsim_freqs']
        self.Nsens = Nsens
        self.Nsim_noise=param['Nsim_noise']
        self.Nts = Nts
        self.Fs = Fs
        kernel     = vsip.create('kaiser_d',(6,1,0))
        self.fir   = vsip.create('fir_d',(kernel, sym, 2*L, 2, state, 0,0))
        self.noise = vsip.create('vview_d',(2*L,mem))
        self.bl_noise = vsip.create('vview_d',(L,mem))
        self.rand  = vsip.create('randstate',(7,1,1,rng))
        self.t = vsip.create('vview_d',(Nts,mem))
        vsip.ramp(0,1.0/Fs,self.t) #vector of sample times
        self.t_dt = vsip.create('vview_d',(Nts,mem))    
        self.m_data = vsip.create('mview_d',(Nsens,Nts,row,mem))
        self.v_data = vsip.rowview(self.m_data,0)
        self.d_t = param['Dsens']/param['c'] #travel time at end-fire between sensors
        self.sim_freqs = param['sim_freqs']
        self.sim_bearings = param['sim_bearings']
        vsip.allDestroy(kernel)

    def __del__(self):
        vsip.destroy(self.fir)
        vsip.allDestroy(self.noise)
        vsip.allDestroy(self.bl_noise)
        vsip.destroy(self.rand)
        vsip.allDestroy(self.t)
        vsip.allDestroy(self.t_dt)
        vsip.destroy(self.v_data)
        vsip.allDestroy(self.m_data)

    def nb_sim(self):
        from numpy import pi,cos
        for i in range(self.Nsim_freqs):
            f=self.sim_freqs[i]
            b=self.d_t * cos(self.sim_bearings[i] * pi/180.0)
            for j in range(self.Nsens):
                dt = float(j) * b
                vsip.add(dt,self.t,self.t_dt)
                vsip.mul(2 * pi * f, self.t_dt, self.t_dt);
                vsip.cos(self.t_dt,self.t_dt)
                vsip.mul(3.0,self.t_dt,self.t_dt)
                VU_rowview(self.v_data,self.m_data,j)
                vsip.add(self.t_dt,self.v_data,self.v_data)
    
    def noise_sim(self):
        from numpy import pi, cos
        d_t=self.d_t * self.Fs #sensor-to-sensor travel time at end-fire in samples
        o_0    = d_t * self.Nsens + 1 # array travel time at end-fire in samples
        a_stp = pi/self.Nsim_noise # angle step
        bl_attr = vsip.getattrib(self.bl_noise)
        for j in range(self.Nsim_noise):
            a_crct = cos(float(j) * a_stp)
            vsip.randn(self.rand,self.noise)
            vsip.firfilt(self.fir,self.noise,self.bl_noise)
            vsip.mul(12.0/float(self.Nsim_noise),self.bl_noise,self.bl_noise)
            vsip.putlength(self.bl_noise,self.Nts);
            for i in range(self.Nsens):
               vsip.putoffset(self.bl_noise,int(o_0 + i * d_t * a_crct))
               VU_rowview(self.v_data,self.m_data,i)
               vsip.add(self.bl_noise,self.v_data,self.v_data)
            vsip.putattrib(self.bl_noise,bl_attr);
        vsip.add(-vsip.meanval(self.m_data),self.m_data,self.m_data);

    def zero(self):
        vsip.fill(0.0,self.m_data)

    def instance(self):
        return self.m_data

File: AppleXcode/KOmegaBFEx.playground/Sources/ts.swift
import Foundation
import vsip

public class TimeSeries{
    public let Fs: Double
    public let c: Double
    public let Dsens: Double
    public let Nsens: Int
   public  let Nsim_noise: Int
   public  let Nsim: Int
   public  let Nts: Int
   public  var fir: OpaquePointer?
   public  var noise: OpaquePointer?
    public var bl_noise: OpaquePointer?
    public var rand: OpaquePointer?
    public var t: OpaquePointer?
    public var t_dt: OpaquePointer?
    public var m_data: OpaquePointer?
    public var v_data: OpaquePointer?
    public var freqs: [Double] = []
    public var bearings: [Double] = []
    public var d_t: Double {
        return self.Dsens/self.c
    }
    public init(param: Param) {
        var tL = 2.0 * param.Fs!
        tL /= Double(param.Nsens!) * param.Dsens!/param.c!
        tL += Double(param.Nts! + 1)
        let L = Int(2 * tL)
        self.Nsim = param.Nsim!
        self.freqs = param.freqs
        self.bearings = param.bearings
        self.Nts = param.Nts!
        self.Nsens = param.Nsens!
        self.c = param.c!
        self.Dsens = param.Dsens!
        self.Nsim_noise=param.Nsim_noise!
        self.Fs = param.Fs!
        
        if let kernel = vsip_vcreate_kaiser_d(6,1,VSIP_MEM_NONE){
            if let t = vsip_fir_create_d(kernel,VSIP_NONSYM, vsip_length(L) , 2 , VSIP_STATE_SAVE,0,VSIP_ALG_TIME){
                self.fir = t
            }
            if let t = vsip_vcreate_d(vsip_length(L),VSIP_MEM_NONE){
                self.noise = t
            }
            if let t = vsip_vcreate_d(vsip_length(Int(tL)),VSIP_MEM_NONE) {
                self.bl_noise = t
            }
            if let t = vsip_randcreate(7,1,1,VSIP_PRNG) {
                self.rand = t
            }
            if let t = vsip_vcreate_d(vsip_length(param.Nts!),VSIP_MEM_NONE) {
                self.t = t
            }
            if let t = vsip_vcreate_d(vsip_length(param.Nts!),VSIP_MEM_NONE) {
                self.t_dt = t
                let step = 1.0 / param.Fs!
                vsip_vramp_d(0.0,step,self.t);
            }
            if let t = vsip_mcreate_d(vsip_length(param.Nsens!),vsip_length(param.Nts!),VSIP_ROW,VSIP_MEM_NONE) {
                self.m_data = t
                if let t = vsip_mrowview_d(self.m_data,0){
                    self.v_data = t
                }
            }
            vsip_valldestroy_d(kernel)
        }
    }
    deinit {
        vsip_fir_destroy_d(self.fir);
        vsip_valldestroy_d(self.noise);
        vsip_valldestroy_d(self.bl_noise);
        vsip_randdestroy(self.rand);
        vsip_valldestroy_d(self.t);
        vsip_valldestroy_d(self.t_dt);
        vsip_vdestroy_d(self.v_data);
        vsip_malldestroy_d(self.m_data);
    }
    public func zero(){
        vsip_mfill_d(0.0,m_data)
    }
    public func rowview(_ i: Int){
        var mattr: vsip_mattr_d = vsip_mattr_d()
        var vattr: vsip_vattr_d = vsip_vattr_d()
        vsip_mgetattrib_d(m_data,&mattr)
        vattr.offset = vsip_offset(Int(mattr.offset) + i * Int(mattr.col_stride))
        vattr.length = mattr.row_length
        vattr.stride = mattr.row_stride
        vsip_vputattrib_d(v_data,&vattr)
    }
    public func nb_sim(){
        let d_t = self.d_t
        for i in 0..<Nsim {
            let f = freqs[i]
            let b = d_t * cos(bearings[i] * Double.pi/180)
            for j in 0..<Nsens {
                let dt = Double(j) * b
                vsip_svadd_d(dt,t,t_dt)
                vsip_svmul_d(2.0 * Double.pi * f,t_dt,t_dt)
                vsip_vcos_d(t_dt,t_dt)
                vsip_svmul_d(3.0,t_dt,t_dt)
                rowview(j);
                vsip_vadd_d(t_dt,v_data,v_data);
            } // next sensor
        } // next bearing
        
    }
    public func noise_sim(){
        let d_t = Fs * self.d_t
        let o_0 = d_t * Double(Nsens) + 1.0
        let a_stp = Double.pi / Double(self.Nsim_noise)
        var bl_attr: vsip_vattr_d = vsip_vattr_d()
        vsip_vgetattrib_d(bl_noise,&bl_attr)
        for j in 0..<Nsim_noise {
            let a_crct = cos(Double(j) * a_stp)
            vsip_vrandn_d(rand,noise);
            vsip_firflt_d(fir,noise,bl_noise);
            vsip_svmul_d(12.0/Double(Nsim_noise),bl_noise,bl_noise);
            vsip_vputlength_d(bl_noise,vsip_length(Nts))
            for i in 0..<Nsens {
                let index = vsip_index(o_0 + Double(i) * d_t * a_crct)
                vsip_vputoffset_d(bl_noise, index)
                rowview(i)
                vsip_vadd_d(bl_noise,v_data,v_data);
            }
            vsip_vputattrib_d(bl_noise,&bl_attr);
        }
        vsip_smadd_d(-vsip_mmeanval_d(m_data),m_data,m_data);        
    }
    public func instance() -> OpaquePointer {
        return self.m_data!
    }
}

